lib/models.py

```diff
@@ -2,13 +2,267 @@
 import openai
 from typing import Any, Callable, List, Optional
 
-class CoT():
+import numpy as np
+import itertools
+
+class ToT():
+    """
+    Creates a ToT wrapper around an LLM. 
+
+    Is called exactly like an LLM, just executes a chain of thought to get answer.
+
+    Based on code from this paper: https://arxiv.org/pdf/2305.10601.pdf
     """
-    Creates a CoT wrapper around an LLM. 
 
-    Is called exactly like an LLM, just excecutes a chain of thought to get answer.
+    num_steps: int = 3               # the depth of the chain of thought tree
+    num_select_sample: int = 2       # how many samples to keep exploring at the level of the tree
+    num_generate_samples: int = 3    # how many samples to generate for each sample of the tree (branching fraction)
+    verbose: bool = False            # weather or not to print out the CoT
+    selection_method: str = 'greedy' # Describes how set of best possible solutions is selected. Can either be 'greedy' or 'sample'.
+    max_num_tries: int = 3           # maximum number of tries to get a score or final answer before it gives up.
+    return_boolean: bool = True      # flag which, when set to true, forces CoT to output a boolean string (either "true" or "false")
+    get_samples_prompt: str = ''     # prompt to get the samples. Requires "Problem" and "Previous_CoT" as placeholders
+    get_scores_prompt: str = ''      # prompt to get the scores. Requiress "CoT" as a placeholder.
+    get_answer_prompt: str = ''      # prompt to get the final answer. Requires "CoT" and "Problem" as a placeholder.
+
+    def __init__(self, llm, **kwargs):
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+        
+        self.llm = llm
+
+        # Check if the neccessary prompts have been loaded into the config
+        if len(self.get_samples_prompt) == 0:
+            raise AssertionError('Sample prompt must be included in CoT configuration')
+        if len(self.get_scores_prompt) == 0:
+            raise AssertionError('Score prompt must be included in CoT configuration')
+        if len(self.get_answer_prompt) == 0:
+            raise AssertionError('Answer prompt must be included in CoT configuration')
+
+    def get_samples(self, problem, previous_CoT):
+        """
+        Takes an input problem, x, and an existing chain of thought, y, 
```

Comment:
Update comments to reflect new variable names (`x` and `y` become `problem` and `previous_CoT`)